pxr/usd/usdSemantics/labelsQuery.h

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright 2024 Pixar
+//
+// Licensed under the Apache License, Version 2.0 (the "Apache License")
+// with the following modification; you may not use this file except in
+// compliance with the Apache License and the following modification to it:
+// Section 6. Trademarks. is deleted and replaced with:
+//
+// 6. Trademarks. This License does not grant permission to use the trade
+//    names, trademarks, service marks, or product names of the Licensor
+//    and its affiliates, except as required to comply with Section 4(c) of
+//    the License and to reproduce the content of the NOTICE file.
+//
+// You may obtain a copy of the Apache License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the Apache License with the above modification is
+// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the Apache License for the specific
+// language governing permissions and limitations under the Apache License.
+//
+
+#ifndef PXR_USDSEMANTICS_LABELSQUERY_H
+#define PXR_USDSEMANTICS_LABELSQUERY_H
+
+#include "pxr/pxr.h"
+
+#include "pxr/usd/usdSemantics/labelsAPI.h"
+#include "pxr/usd/usd/timeCode.h"
+#include "pxr/base/gf/interval.h"
+#include "pxr/base/tf/token.h"
+
+#include <shared_mutex>
+#include <set>
+#include <unordered_map>
+#include <variant>
+
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+/// The `UsdSemanticsLabelsQuery` can be used to query a prim's
+/// labels for a specified taxonomy and time from the `UsdSemanticsLabelsAPI`.
+/// Time may be an individual time code or an interval.
+///
+/// This utility requires that all prims are on the same stage.
+///
+/// The query caches certain reads and computations and should be discarded
+/// when the state of the stage changes.
+class UsdSemanticsLabelsQuery
+{
+public:
+    UsdSemanticsLabelsQuery(const UsdSemanticsLabelsQuery&) = delete;
+    UsdSemanticsLabelsQuery& operator=(
+        const UsdSemanticsLabelsQuery&) = delete;
+
+    /// Constructs a query for a `taxonomy` at a single `timeCode`.
+    /// 
+    /// Requires that the `taxonomy` must not be empty.
+    USDSEMANTICS_API UsdSemanticsLabelsQuery(const TfToken& taxonomy,
+                                             UsdTimeCode timeCode);
+
+    /// Construct a query for a `taxonomy` over an `interval`.
+    ///
+    /// Requires that neither the `interval` nor `taxonomy` to be empty.
+    ///
+    /// \warning Finite minimum values of the interval will return the same
+    /// result regardless of closed or open state due to held interpolation
+    /// semantics and Zeno's paradox.
+    USDSEMANTICS_API UsdSemanticsLabelsQuery(const TfToken& taxonomy,
+                                             const GfInterval& interval);
+
+    /// Compute the union of values for `semantics:labels:<taxonomy>` in the
+    /// specified interval.
```

Comment:
For this and other methods that return an array of labels, please specify whether there is a stable ordering of the returned values, or client must expect them to be disordered.

For this method specifically, it WBN to just reinforce the convention that "Direct" means this prim's contributions only.